106-auto-rebalance-on-imbalanced-clusters.md

```diff
@@ -0,0 +1,471 @@
+
+#### AnomalyDetectorNotifier
+
+Cruise Control provides the `AnomalyNotifier` interface, which has multiple abstract methods on what to do if certain anomalies are detected.
+Some of those methods are:`onGoalViolation()`, `onBrokerFailure()`, `onDiskFailure`, `alert()` etc.
+The `AnomalyDetectorNotifier` will be based of the `AnomalyNotifier` class.
+The anomalies with smaller priority value and detected time will be considered priority and resolved first.
+In case the anomaly is unfixable due to issues like  lack of physical hardware (e.g. insufficient number of racks to satisfy rack awareness, insufficient number of brokers to satisfy ReplicaCapacityGoal, or insufficient number of resources to satisfy resource capacity goals), then they will be ignored.
+The `AnomalyDetectorNotifier` will override all the abstract methods provided by the `AnomalyNotifier` interface.
+We will add an `alert` method which will alert the operator whenever an anomaly is detected by the operator.
+Upon detection of an anomaly, the notifier would create a configmap with name set as `goal-violation` followed by `anomalyId`.
+
+The ConfigMap would look like this:
+
+```yaml
+kind: ConfigMap
+apiVersion: v1
+metadata:
+  name: goal-violation-<anomalyID>
+data:
+  anomalyType: GOAL_VIOLATION
+  anomalyId: "<anomaly-id>"
+  detectionTime: "<time-of-detection>"
+# ...
+```
+
+The operator will then check if any configmap with prefix `goal-violation` is created or not, if it finds one created then operator will trigger the rebalance.
+Separate configmaps would be created for every goal violation such that on completion of the rebalance we can remove the particular configmap.
+The advantages of using a separate configmap for every anomaly are:
+1. Every anomaly would have a separate place to put their data in.
+2. Better readability.  
+3. Improves scope for future improvement when dealing with different violation, for example topic or metrics related violations.
+
+#### Using other notifiers when using `skew` mode
+
+Users cannot configure the notifier if they are utilising the auto-rebalance on imbalanced cluster.
+This is because the operator is using our custom notifier for getting alerts about goal violations. 
+If the users try to override the notifier while the `skew` mode is enabled, the auto-rebalance `skew` configuration then the operator would throw errors in the auto-rebalance status field
+
+If the users really want to have their own way of dealing with the imbalanced clusters then they can just disable auto-rebalance in `skew` mode and use their own notifier.
+
+#### What happens if some unfixable goal violation happens
+In case, there is an unfixable goal violation then the notifier would simply ignore that anomaly and prompt the user about the unfixable violation in the auto-rebalancing status section.
```

Comment:
```suggestion
If there is an unfixable goal violation, then the notifier would simply ignore that anomaly and prompt the user about the unfixable violation in the auto-rebalancing status section.
```

Reply:
Would this appear in the Kafka CR? I agree with Paolo that you need an example here.